File: implementation/swift/GameEventScriptConformance/Sources/GameEventScriptConformance/ConformanceMarkdownParser.swift
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

import Foundation

/// The fileless, synchronous parser for strict Conformance Markdown V1.
public enum ConformanceMarkdownParser {
    /// Parses and validates a complete Markdown conformance document under resource limits. The byte overload rejects
    /// malformed UTF-8; no partial model is returned.
    ///
    /// - Throws: `ConformanceParseError` with stable code and source range.
    public static func parse(_ text: String, limits: ConformanceParserLimits = .default) throws -> ConformanceDocument { try parse(Array(text.utf8), limits: limits) }

    /// Parses and validates a complete Markdown conformance document under resource limits. The byte overload rejects
    /// malformed UTF-8; no partial model is returned.
    ///
    /// - Throws: `ConformanceParseError` with stable code and source range.
    public static func parse(_ bytes: [UInt8], limits: ConformanceParserLimits = .default) throws -> ConformanceDocument {
        guard limits.maxDocumentBytes > 0, bytes.count <= limits.maxDocumentBytes else { throw ConformanceParseError("conformance.yaml.limitExceeded", "The document exceeds maxDocumentBytes.") }
        let start = bytes.starts(with: [0xef, 0xbb, 0xbf]) ? 3 : 0
        guard !bytes.contains(0), String(bytes: bytes.dropFirst(start), encoding: .utf8) != nil else { throw ConformanceParseError("conformance.markdown.invalidUtf8", "Conformance documents must be strict UTF-8 without NUL bytes.") }
        var lines: [ConformanceLine] = []
        var begin = start
        var index = start
        while index < bytes.count {
            if bytes[index] != 10 && bytes[index] != 13 {
                index += 1
                continue
            }
            lines.append(.init(String(decoding: bytes[begin..<index], as: UTF8.self), number: lines.count + 1, byteOffset: begin - start))
            let ending = bytes[index] == 13 && index + 1 < bytes.count && bytes[index + 1] == 10 ? 2 : 1
            index += ending
            begin = index
        }
        if begin < bytes.count || lines.isEmpty { lines.append(.init(String(decoding: bytes[begin...], as: UTF8.self), number: lines.count + 1, byteOffset: begin - start)) }
        let syntax = try ConformanceMarkdownScanner.scan(lines, limits: limits)
        return try ConformanceSchema.bind(syntax, bytes: bytes, limits: limits)
    }
}

internal struct ConformanceMarkdownBlock {
    let info: String
    let lines: [ConformanceLine]
    let range: ConformanceSourceRange
    let payloadRange: ConformanceSourceRange
    var payload: String { lines.map(\.text).joined(separator: "\n") }
}

internal struct ConformanceMarkdownStep {
    let id: String
    let receive: String
    let pump: String
    let budget: String
    let range: ConformanceSourceRange
}

internal struct ConformanceMarkdownCase {
    let title: String
    var range: ConformanceSourceRange
    var yaml: [ConformanceMarkdownBlock] = []
    var sources: [ConformanceMarkdownBlock] = []
    var assembler: ConformanceMarkdownBlock?
    var steps: [ConformanceMarkdownStep] = []
    var hasSteps = false
    var stepsTableRange: ConformanceSourceRange?
}

internal struct ConformanceMarkdownSyntax {
    let frontmatter: [ConformanceLine]
    let frontmatterRange: ConformanceSourceRange
    let cases: [ConformanceMarkdownCase]
}

internal enum ConformanceMarkdownScanner {
    static func scan(_ lines: [ConformanceLine], limits: ConformanceParserLimits) throws -> ConformanceMarkdownSyntax {
        guard lines.first?.text == "---" else { throw failure("missingFrontmatter", "A document must begin with YAML frontmatter.", lines.first?.range()) }
        guard let close = lines.dropFirst().firstIndex(where: { $0.text == "---" }) else { throw failure("unterminatedFrontmatter", "Frontmatter must end with '---'.", lines[0].range()) }
        var cases: [ConformanceMarkdownCase] = []
        var fixtures = false
        var fixtureSeen = false
        var i = close + 1
        while i < lines.count {
            let line = lines[i]
            if line.text.hasPrefix("```") {
                let info = String(line.text.dropFirst(3))
                let yaml = !fixtures && !cases.isEmpty && info == "yaml"
                let semantic = ["ges", "gesa"].contains(info) || yaml
                let suspicious = info.hasPrefix("ges") || (!fixtures && !cases.isEmpty && info.hasPrefix("yaml "))
                var end = i + 1
                while end < lines.count && lines[end].text != "```" {
                    if semantic && lines[end].text.hasPrefix("```") { throw failure("unknownSemanticFence", "Semantic fences cannot be nested.", lines[end].range()) }
                    end += 1
                }
                if end == lines.count {
                    if semantic || suspicious { throw failure("unterminatedFence", "A recognized fence is not terminated.", line.range()) }
                    i += 1
                    continue
                }
                if !fixtures {
                    if cases.isEmpty && semantic { throw schema("invalidCardinality", "Semantic fences belong inside a test.", line.range()) }
                    if !cases.isEmpty {
                        if !semantic && suspicious { throw failure("unknownSemanticFence", "Unknown semantic fence '\(info)'.", line.range(3)) }
                        let payloadRange = end == i + 1 ? lines[end].range(0, 0) : spanning(lines[i + 1].range(), through: lines[end - 1])
                        let block = ConformanceMarkdownBlock(info: info, lines: Array(lines[(i + 1)..<end]), range: spanning(line.range(), through: lines[end]), payloadRange: payloadRange)
                        if info == "yaml" { cases[cases.count - 1].yaml.append(block) }
                        if info == "ges" { cases[cases.count - 1].sources.append(block) }
                        if info == "gesa" {
                            guard cases[cases.count - 1].assembler == nil else { throw schema("invalidCardinality", "A test accepts at most one gesa block.", line.range()) }
                            cases[cases.count - 1].assembler = block
                        }
                    }
                }
                i = end + 1
                continue
            }
            if line.text.hasPrefix("## Test:") {
                guard line.text.hasPrefix("## Test: ") else { throw failure("invalidTestHeading", "Test headings use '## Test: Display title'.", line.range()) }
                let title = ConformanceYamlParser.trim(String(line.text.dropFirst(9)))
                guard !title.isEmpty else { throw failure("invalidTestHeading", "A test display title cannot be empty.", line.range()) }
                guard limits.maxTests > 0, cases.count < limits.maxTests else { throw ConformanceParseError("conformance.yaml.limitExceeded", "Document exceeds maxTests.", line.range()) }
                if !cases.isEmpty { cases[cases.count - 1].range = spanning(cases[cases.count - 1].range, through: lines[i - 1]) }
                cases.append(.init(title: title, range: line.range()))
                fixtures = false
            } else if line.text == "## Fixtures" {
                guard cases.isEmpty, !fixtureSeen else { throw schema("invalidCardinality", "Fixtures may occur once, before the first test.", line.range()) }
                fixtureSeen = true
                fixtures = true
            } else if !cases.isEmpty && line.text == "### Steps" {
                guard !cases[cases.count - 1].hasSteps else { throw failure("invalidStepsTable", "A test can contain only one Steps table.", line.range()) }
                i += 1
                while i < lines.count && ConformanceYamlParser.trim(lines[i].text).isEmpty { i += 1 }
                guard i + 1 < lines.count, lines[i].text == "| step | receive | pump | budget |", lines[i + 1].text == "| --- | --- | --- | --- |" else {
                    throw failure("invalidStepsTable", "The Steps table header or delimiter is invalid.", line.range())
                }
                i += 2
                while i < lines.count && lines[i].text.hasPrefix("|") {
                    let row = lines[i]
                    guard row.text.hasSuffix("|"), !row.text.contains("\\"), !row.text.contains("`") else { throw failure("invalidStepsTable", "Steps rows require unescaped plain cells.", row.range()) }
                    let cells = row.text.dropFirst().dropLast().split(separator: "|", omittingEmptySubsequences: false).map { ConformanceYamlParser.trim(String($0)) }
                    guard cells.count == 4 else { throw failure("invalidStepsTable", "A Steps row requires four cells.", row.range()) }
                    guard limits.maxStepsPerTest > 0, cases[cases.count - 1].steps.count < limits.maxStepsPerTest else { throw ConformanceParseError("conformance.yaml.limitExceeded", "Steps exceed maxStepsPerTest.", row.range()) }
                    cases[cases.count - 1].steps.append(.init(id: cells[0], receive: cells[1], pump: cells[2], budget: cells[3], range: row.range()))
                    i += 1
                }
                guard !cases[cases.count - 1].steps.isEmpty else { throw failure("invalidStepsTable", "Steps requires at least one row.", line.range()) }
                cases[cases.count - 1].hasSteps = true
                cases[cases.count - 1].stepsTableRange = spanning(line.range(), through: lines[i - 1])
                continue
            }
            i += 1
        }
        guard !cases.isEmpty else { throw schema("invalidCardinality", "A suite requires at least one test.", lines[close].range()) }
        cases[cases.count - 1].range = spanning(cases[cases.count - 1].range, through: lines[lines.count - 1])
        return .init(frontmatter: Array(lines[1..<close]), frontmatterRange: spanning(lines[0].range(), through: lines[close]), cases: cases)
    }

    private static func spanning(_ first: ConformanceSourceRange, through last: ConformanceLine) -> ConformanceSourceRange {
        let lastRange = last.range()
        return .init(byteOffset: first.byteOffset, byteLength: lastRange.byteOffset + lastRange.byteLength - first.byteOffset, line: first.line, column: first.column, endLine: lastRange.endLine, endColumn: lastRange.endColumn)
    }

    private static func failure(_ code: String, _ message: String, _ range: ConformanceSourceRange?) -> ConformanceParseError { .init("conformance.markdown.\(code)", message, range ?? .init()) }

    private static func schema(_ code: String, _ message: String, _ range: ConformanceSourceRange) -> ConformanceParseError { .init("conformance.schema.\(code)", message, range) }
}

File: implementation/csharp/GameEventScript.Conformance/src/ConformanceMarkdownParser.cs
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections.Generic;
using System.Text;

namespace GameEventScript.Conformance;

/// <summary>
/// Parses the strict, portable Conformance Markdown V1 subset into an immutable normalized model.
/// </summary>
public static class ConformanceMarkdownParser
{
    /// <summary>
    /// Parses strict UTF-8 Conformance Markdown using the default bounded limits.
    /// </summary>
    /// <param name="utf8Bytes">The complete UTF-8 document bytes.</param>
    /// <returns>The normalized conformance document.</returns>
    public static ConformanceDocument Parse(byte[] utf8Bytes)
        => Parse(utf8Bytes, ConformanceParserLimits.Default);

    /// <summary>
    /// Parses strict UTF-8 Conformance Markdown using explicit bounded limits.
    /// </summary>
    /// <param name="utf8Bytes">The complete UTF-8 document bytes.</param>
    /// <param name="limits">The parser resource limits.</param>
    /// <returns>The normalized conformance document with retained UTF-8 source ranges.</returns>
    /// <exception cref="ConformanceParseException">Thrown with stable diagnostics when syntax, schema, encoding, or a resource limit is invalid.</exception>
    public static ConformanceDocument Parse(byte[] utf8Bytes, ConformanceParserLimits limits)
    {
        if (utf8Bytes is null) throw new ArgumentNullException(nameof(utf8Bytes));
        if (limits is null) throw new ArgumentNullException(nameof(limits));
        try
        {
            var text = ConformanceText.Decode(utf8Bytes, limits);
            var syntax = MarkdownScanner.Scan(text, limits);
            return ConformanceSchemaBinder.Bind(text, syntax, limits);
        }
        catch (ConformanceFailure failure)
        {
            throw new ConformanceParseException(new[] { failure.Diagnostic });
        }
    }

    /// <summary>
    /// Encodes and parses a .NET string as Conformance Markdown using the default limits.
    /// </summary>
    /// <param name="text">The complete Markdown text; unpaired UTF-16 surrogates are rejected.</param>
    /// <returns>The normalized conformance document.</returns>
    public static ConformanceDocument Parse(string text)
        => Parse(text, ConformanceParserLimits.Default);

    /// <summary>
    /// Encodes and parses a .NET string as Conformance Markdown using explicit limits.
    /// </summary>
    /// <param name="text">The complete Markdown text; unpaired UTF-16 surrogates are rejected.</param>
    /// <param name="limits">The parser resource limits.</param>
    /// <returns>The normalized conformance document with retained UTF-8 source ranges.</returns>
    /// <exception cref="ConformanceParseException">Thrown with stable diagnostics when syntax, schema, encoding, or a resource limit is invalid.</exception>
    public static ConformanceDocument Parse(string text, ConformanceParserLimits limits)
    {
        if (text is null) throw new ArgumentNullException(nameof(text));
        try
        {
            return Parse(new UTF8Encoding(false, true).GetBytes(text), limits);
        }
        catch (EncoderFallbackException)
        {
            throw new ConformanceParseException(new[]
            {
                new ConformanceDiagnostic(ConformanceDiagnosticCodes.InvalidUtf8, "The input contains an unpaired UTF-16 surrogate.", new ConformanceSourceRange(0, 0, 1, 1, 1, 1))
            });
        }
    }
}

internal sealed class MarkdownSyntax
{
    internal MarkdownSyntax(List<ConformanceLine> frontmatter, int frontmatterStart, int frontmatterEnd, List<MarkdownCaseSyntax> cases)
    {
        Frontmatter = frontmatter;
        FrontmatterStart = frontmatterStart;
        FrontmatterEnd = frontmatterEnd;
        Cases = cases;
    }

    internal List<ConformanceLine> Frontmatter { get; }
    internal int FrontmatterStart { get; }
    internal int FrontmatterEnd { get; }
    internal List<MarkdownCaseSyntax> Cases { get; }
}

internal sealed class MarkdownCaseSyntax
{
    internal MarkdownCaseSyntax(string title, int startLineIndex)
    {
        Title = title;
        StartLineIndex = startLineIndex;
    }

    internal string Title { get; }
    internal int StartLineIndex { get; }
    internal int EndLineIndex { get; set; }
    internal MarkdownBlockSyntax? CaseBlock { get; set; }
    internal MarkdownBlockSyntax? ExpectBlock { get; set; }
    internal MarkdownBlockSyntax? AssemblerBlock { get; set; }
    internal List<MarkdownBlockSyntax> YamlBlocks { get; } = new();
    internal List<MarkdownBlockSyntax> Sources { get; } = new();
    internal List<MarkdownStepSyntax> Steps { get; } = new();
    internal bool HasStepsTable { get; set; }
    internal ConformanceSourceRange? StepsTableRange { get; set; }
}

internal sealed class MarkdownBlockSyntax
{
    internal MarkdownBlockSyntax(string info, List<ConformanceLine> payloadLines, ConformanceSourceRange blockRange, ConformanceSourceRange payloadRange)
    {
        Info = info;
        PayloadLines = payloadLines;
        BlockRange = blockRange;
        PayloadRange = payloadRange;
    }

    internal string Info { get; }
    internal List<ConformanceLine> PayloadLines { get; }
    internal ConformanceSourceRange BlockRange { get; }
    internal ConformanceSourceRange PayloadRange { get; }
    internal string Payload => ConformanceText.JoinPayload(PayloadLines);
}

internal sealed class MarkdownStepSyntax
{
    internal MarkdownStepSyntax(string id, string receive, string pump, string budget, ConformanceSourceRange range)
    {
        Id = id; Receive = receive; Pump = pump; Budget = budget; Range = range;
    }
    internal string Id { get; }
    internal string Receive { get; }
    internal string Pump { get; }
    internal string Budget { get; }
    internal ConformanceSourceRange Range { get; }
}

internal static class MarkdownScanner
{
    private static readonly HashSet<string> RecognizedFences = new(StringComparer.Ordinal)
    {
        "ges", "gesa"
    };

    internal static MarkdownSyntax Scan(ConformanceText text, ConformanceParserLimits limits)
    {
        var lines = text.Lines;
        if (lines.Count == 0 || lines[0].Text != "---")
            throw Fail(ConformanceDiagnosticCodes.MissingFrontmatter, "The document must start with YAML frontmatter.", lines.Count == 0 ? Start() : lines[0].Range());
        var close = -1;
        for (var index = 1; index < lines.Count; index++)
        {
            if (lines[index].Text == "---") { close = index; break; }
        }
        if (close < 0) throw Fail(ConformanceDiagnosticCodes.UnterminatedFrontmatter, "The YAML frontmatter is not terminated by '---'.", lines[0].Range());

        var frontmatter = lines.GetRange(1, close - 1);
        var cases = new List<MarkdownCaseSyntax>();
        MarkdownCaseSyntax? current = null;
        var fixtureMode = false;
        var fixtureSeen = false;
        for (var index = close + 1; index < lines.Count; index++)
        {
            var line = lines[index];
            if (line.Text.StartsWith("```", StringComparison.Ordinal))
            {
                var info = line.Text.Length == 3 ? string.Empty : line.Text.Substring(3);
                var yamlInTest = !fixtureMode && current is not null && info == "yaml";
                var semantic = RecognizedFences.Contains(info) || yamlInTest;
                var suspicious = info.StartsWith("ges", StringComparison.Ordinal)
                    || (!fixtureMode && current is not null && info.StartsWith("yaml ", StringComparison.Ordinal));
                var end = FindFenceEnd(lines, index + 1, semantic);
                if (end < 0 && (semantic || suspicious)) throw Fail(ConformanceDiagnosticCodes.UnterminatedFence, "A recognized semantic fence is not terminated.", line.Range());
                if (end < 0) continue;
                if (!fixtureMode && current is null && semantic)
                    throw Fail(ConformanceDiagnosticCodes.SchemaInvalidCardinality, "A semantic fence must occur inside a test.", line.Range());
                if (!fixtureMode && current is not null)
                {
                    if (!semantic && suspicious) throw Fail(ConformanceDiagnosticCodes.UnknownSemanticFence, $"Unknown semantic fence '{info}'.", line.Range(3));
                    if (semantic)
                    {
                        var payload = lines.GetRange(index + 1, end - index - 1);
                        var block = new MarkdownBlockSyntax(info, payload, text.LinesRange(index, end), ConformanceText.PayloadRange(payload, line));
                        AddBlock(current, block);
                    }
                }
                index = end;
                continue;
            }

            if (line.Text.StartsWith("## Test:", StringComparison.Ordinal))
            {
                if (!line.Text.StartsWith("## Test: ", StringComparison.Ordinal)) throw Fail(ConformanceDiagnosticCodes.InvalidTestHeading, "A test heading must use '## Test: Display title'.", line.Range());
                var title = line.Text.Substring(9).Trim(' ', '\t');
                if (title.Length == 0) throw Fail(ConformanceDiagnosticCodes.InvalidTestHeading, "A test heading requires a display title.", line.Range());
                if (current is not null) current.EndLineIndex = index - 1;
                if (limits.MaxTests <= 0 || cases.Count >= limits.MaxTests) throw Fail(ConformanceDiagnosticCodes.YamlLimitExceeded, "The document exceeds MaxTests.", line.Range());
                current = new MarkdownCaseSyntax(title, index);
                cases.Add(current);
                fixtureMode = false;
                continue;
            }

            if (line.Text == "## Fixtures")
            {
                if (current is not null || fixtureSeen) throw Fail(ConformanceDiagnosticCodes.SchemaInvalidCardinality, "The Fixtures section may occur once and only before the first test.", line.Range());
                fixtureSeen = true;
                fixtureMode = true;
                continue;
            }

            if (current is not null && line.Text == "### Steps")
            {
                if (current.HasStepsTable) throw Fail(ConformanceDiagnosticCodes.InvalidStepsTable, "A test may contain at most one Steps table.", line.Range());
                index = ParseSteps(lines, index, current, limits);
            }
        }

        if (current is not null) current.EndLineIndex = Math.Max(current.StartLineIndex, lines.Count - 1);
        if (cases.Count == 0) throw Fail(ConformanceDiagnosticCodes.SchemaInvalidCardinality, "A suite must contain at least one test.", lines[close].Range());
        return new MarkdownSyntax(frontmatter, 0, close, cases);
    }

    private static int FindFenceEnd(IReadOnlyList<ConformanceLine> lines, int start, bool rejectNested)
    {
        for (var index = start; index < lines.Count; index++)
        {
            if (lines[index].Text == "```") return index;
            if (rejectNested && lines[index].Text.StartsWith("```", StringComparison.Ordinal))
                throw Fail(ConformanceDiagnosticCodes.UnknownSemanticFence, "Semantic fences cannot be nested.", lines[index].Range());
        }
        return -1;
    }

    private static void AddBlock(MarkdownCaseSyntax test, MarkdownBlockSyntax block)
    {
        switch (block.Info)
        {
            case "yaml":
                test.YamlBlocks.Add(block);
                break;
            case "gesa":
                if (test.AssemblerBlock is not null) throw Fail(ConformanceDiagnosticCodes.SchemaInvalidCardinality, "A test may contain at most one gesa block.", block.BlockRange);
                test.AssemblerBlock = block;
                break;
            case "ges":
                test.Sources.Add(block);
                break;
        }
    }

    private static int ParseSteps(IReadOnlyList<ConformanceLine> lines, int headingIndex, MarkdownCaseSyntax test, ConformanceParserLimits limits)
    {
        var index = headingIndex + 1;
        while (index < lines.Count && string.IsNullOrWhiteSpace(lines[index].Text)) index++;
        if (index + 1 >= lines.Count || lines[index].Text != "| step | receive | pump | budget |" || lines[index + 1].Text != "| --- | --- | --- | --- |")
            throw Fail(ConformanceDiagnosticCodes.InvalidStepsTable, "The Steps table header or delimiter is invalid.", index < lines.Count ? lines[index].Range() : Start());
        index += 2;
        while (index < lines.Count && lines[index].Text.Length > 0 && lines[index].Text[0] == '|')
        {
            if (limits.MaxStepsPerTest <= 0 || test.Steps.Count >= limits.MaxStepsPerTest) throw Fail(ConformanceDiagnosticCodes.YamlLimitExceeded, "The Steps table exceeds MaxStepsPerTest.", lines[index].Range());
            var cells = SplitTableRow(lines[index]);
            if (cells.Count != 4) throw Fail(ConformanceDiagnosticCodes.InvalidStepsTable, "Each Steps row must have exactly four cells.", lines[index].Range());
            test.Steps.Add(new MarkdownStepSyntax(cells[0], cells[1], cells[2], cells[3], lines[index].Range()));
            index++;
        }
        if (test.Steps.Count == 0) throw Fail(ConformanceDiagnosticCodes.InvalidStepsTable, "A Steps table must contain at least one row.", lines[index - 1].Range());
        test.HasStepsTable = true;
        var first = lines[headingIndex];
        var last = lines[index - 1];
        test.StepsTableRange = new ConformanceSourceRange(first.ByteOffset, last.ByteOffset + last.ByteLength - first.ByteOffset, first.LineNumber, 1, last.LineNumber, last.Text.Length + 1);
        return index - 1;
    }

    private static List<string> SplitTableRow(ConformanceLine line)
    {
        if (line.Text.Length < 2 || line.Text[^1] != '|') throw Fail(ConformanceDiagnosticCodes.InvalidStepsTable, "A Steps row must start and end with '|'.", line.Range());
        var content = line.Text.Substring(1, line.Text.Length - 2);
        var raw = content.Split('|');
        var result = new List<string>(raw.Length);
        foreach (var cell in raw)
        {
            if (cell.IndexOf('\\') >= 0 || cell.IndexOf('`') >= 0) throw Fail(ConformanceDiagnosticCodes.InvalidStepsTable, "Steps cells do not support escapes or inline Markdown.", line.Range());
            result.Add(cell.Trim(' ', '\t'));
        }
        return result;
    }

    private static ConformanceFailure Fail(string code, string message, ConformanceSourceRange range) => new(code, message, range);
    private static ConformanceSourceRange Start() => new(0, 0, 1, 1, 1, 1);
}
